File: GNATS_Standalone/sample/sample/DEMO_Aircraft_Functions_prod2.0.py
# NATS sample
#
# Optimal Synthesis Inc.
#
# Oliver Chen
# 03.11.2020
#
# Demo of aircraft-related functions.
#
# Users can learn how to obtain aircraft instance, show related aircraft information, start/pause/resume simulation and output result trajectory file.

from GNATS_Python_Header_standalone import *

import datetime

planned_dirname = ""
output_filename = ""

simulationInterface.clear_trajectory()

environmentInterface.load_rap(DIR_share + "/tg/rap")
aircraftInterface.load_aircraft(DIR_share + "/tg/trx/TRX_DEMO_WSSS_RJAA_geo.trx", DIR_share + "/tg/trx/TRX_DEMO_WSSS_RJAA_mfl.trx")

if not(aircraftInterface is None):
	aircraftIdArray_withinZone = aircraftInterface.getAircraftIds(float(30.0), float(35.0), float(-115.0), float(-90.0), float(-1.0), float(-1.0))
	if (not(aircraftIdArray_withinZone is None) and (len(aircraftIdArray_withinZone) > 0)) :
		i = 0
		for i in range(0, len(aircraftIdArray_withinZone)):
			curAcId = aircraftIdArray_withinZone[i]
			print("Aircraft id in selected zone = ", curAcId)
	
	print("****************************************")
	
	curAircraft = aircraftInterface.select_aircraft("SQ12")
	if not(curAircraft is None):
		airborne_flight_plan_waypoint_name_array = curAircraft.getFlight_plan_waypoint_name_array()
		for j in range(0, len(airborne_flight_plan_waypoint_name_array)) :
			print("SQ12 flight plan waypoint name = ", airborne_flight_plan_waypoint_name_array[j])
		print("")
		
		curAircraft.delay_departure(100) # Delay the departure time for 100 sec

simulationInterface.setupSimulation(43000, 10)

# Start the simulation for 1000 secs then pause
simulationInterface.start(1000)

while True:
	runtime_sim_status = simulationInterface.get_runtime_sim_status()
	#print("Time:", datetime.now(), ", simulation continuing.... ")
	if (runtime_sim_status == GNATS_SIMULATION_STATUS_PAUSE) :
		break
	else:
		time.sleep(1)

curAircraft = aircraftInterface.select_aircraft("SQ12")
if not(curAircraft is None):
	print("****************************************")
	print("SQ12 (pausing at", simulationInterface.get_curr_sim_time(), "sec), latitude = ", curAircraft.getLatitude_deg(), ", longitude = ", curAircraft.getLongitude_deg(), ", altitude = ", curAircraft.getAltitude_ft())
	print("****************************************")

# Resume and continue the simulation
simulationInterface.resume()

# Use a while loop to constantly check simulation status.  When the simulation finishes, continue to output the trajectory data
while True:
	runtime_sim_status = simulationInterface.get_runtime_sim_status()
	#print("Time:", datetime.datetime.now(), ", simulation continuing.... ")
	if (runtime_sim_status == GNATS_SIMULATION_STATUS_ENDED) :
		break
	else:
		time.sleep(1)

millis = int(round(time.time() * 1000))
print("Outputting trajectory data.  Please wait....")

planned_dirname = os.path.splitext(os.path.basename(__file__))[0] + "_" + str(millis)
output_filename = planned_dirname + ".csv"

# Output the trajectory result file
simulationInterface.write_trajectories(output_filename)

aircraftInterface.release_aircraft()
environmentInterface.release_rap()


# Stop GNATS Standalone environment
gnatsStandalone.stop()

# =========================================================

# The following statements read the result trajectory files and display plotting.

# Create temp directory and copy the result trajectory file into it
planned_dirname = "tmp_" + planned_dirname
os.makedirs("../GNATS_Standalone/" + planned_dirname)
copyfile("../GNATS_Standalone/" + output_filename, "../GNATS_Standalone/" + planned_dirname + "/" + output_filename)

post_process = pp.PostProcessor(file_path = "../GNATS_Standalone/" + planned_dirname,
                                ac_name = 'SQ12');
post_process.plotSingleAircraftTrajectory();

# Delete temp directory
print("Deleting directory:", planned_dirname)
rmtree("../GNATS_Standalone/" + planned_dirname) 

shutdownJVM()

File: GNATS_Standalone/sample/sample/DEMO_Aircraft_Functions_prod2_0.m
% GNATS sample
%
% Optimal Synthesis Inc.
%
% Oliver Chen
% 03.12.2020
%
% Demo of aircraft-related functions.
%
% Users can learn how to obtain aircraft instance, show related aircraft information, start/pause/resume simulation and output result trajectory file.

% Run header module
run('GNATS_MATLAB_Header_standalone.m');

if not(isempty(simulationInterface))
    simulationInterface.clear_trajectory();

    environmentInterface.load_rap(strcat(DIR_share, '/tg/rap'));
    
    aircraftInterface.load_aircraft(strcat(DIR_share, '/tg/trx/TRX_DEMO_WSSS_RJAA_geo.trx'), strcat(DIR_share, '/tg/trx/TRX_DEMO_WSSS_RJAA_mfl.trx'));
    
    aircraftIdArray_withinZone = aircraftInterface.getAircraftIds(30.0, 35.0, -115.0, -90.0, -1.0, -1.0);
    if ((~isempty(aircraftIdArray_withinZone)) & (size(aircraftIdArray_withinZone) > 0))
        for i = 1: size(aircraftIdArray_withinZone)
            curAcId = aircraftIdArray_withinZone(i);

            fprintf('Aircraft id in selected zone = %s, time = %ld\n', char(curAcId), round(tic * 1000));
        end
    end
    
    disp('****************************************');

    curAircraft = aircraftInterface.select_aircraft('SQ12');
    if ~isempty(curAircraft)
        airborne_flight_plan_waypoint_name_array = curAircraft.getFlight_plan_waypoint_name_array();

        for i = 1: (size(airborne_flight_plan_waypoint_name_array))
            fprintf('SQ12 airborne flight plan waypoint name = %s\n', char(airborne_flight_plan_waypoint_name_array(i, 1)));
        end
        
        curAircraft.delay_departure(100); % Delay the departure time for 100 sec
    else
        fprintf('Aircraft SQ12 not found');
    end
    
    simulationInterface.setupSimulation(10000, 10);

    simulationInterface.start(1000);

    while true
        runtime_sim_status = simulationInterface.get_runtime_sim_status();
        if (runtime_sim_status == GNATS_SIMULATION_STATUS_PAUSE)
            break;
        else
            pause(1);
        end
    end
    
    curAircraft = aircraftInterface.select_aircraft('SQ12');
    if not(isempty(curAircraft))
        disp('****************************************');
        fprintf('SQ12 (pausing at %f', simulationInterface.get_curr_sim_time());
        fprintf(' sec, latitude = %f', curAircraft.getLatitude_deg());
        fprintf(', longitude = %f', curAircraft.getLongitude_deg());
        fprintf(', altitude = %f\n', curAircraft.getAltitude_ft());
        disp('****************************************');
    end

	simulationInterface.resume();

    % Use a while loop to constantly check simulation status.  When the simulation finishes, continue to output the trajectory data
    while true
        runtime_sim_status = simulationInterface.get_runtime_sim_status();
        if (runtime_sim_status == GNATS_SIMULATION_STATUS_ENDED)
            break;
        else
            pause(1);
        end
    end

    % Format epoch time string
    millis = datestr(now, 'yyyymmdd HHMMSS');
    InputDate = datenum(millis, 'yyyymmdd HHMMSS');
    UnixOrigin = datenum('19700101 000000', 'yyyymmdd HHMMSS');
    EpochSecond = round((InputDate-UnixOrigin)*86400000);

	S = dbstack();
    cur_filename = char(S(1).file);
    strIndexArray = strfind(cur_filename, '.m');

    disp('Outputting trajectory data.  Please wait....');
    fileName = sprintf('%s_%s.csv', cur_filename(1: strIndexArray(1)-1), num2str(EpochSecond));
    % Output the trajectory result file
    simulationInterface.write_trajectories(fileName);
    
    aircraftInterface.release_aircraft();
    environmentInterface.release_rap();
end

% Stop GNATS Standalone environment
gnatsStandalone.stop();

